File: app/src/main/java/com/cxwl/shawn/wuzhishan/decision/fragment/PdfListFragment.java
package com.cxwl.shawn.wuzhishan.decision.fragment;

import android.app.Fragment;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.ListView;
import android.widget.TextView;

import com.cxwl.shawn.wuzhishan.decision.R;
import com.cxwl.shawn.wuzhishan.decision.activity.PDFActivity;
import com.cxwl.shawn.wuzhishan.decision.adapter.PDFListAdapter;
import com.cxwl.shawn.wuzhishan.decision.common.CONST;
import com.cxwl.shawn.wuzhishan.decision.dto.ColumnData;
import com.cxwl.shawn.wuzhishan.decision.util.OkHttpUtil;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Request;
import okhttp3.Response;

/**
 * pdf文档列表
 * @author shawn_sun
 *
 */

public class PdfListFragment extends Fragment {
	
	private ListView listView = null;
	private PDFListAdapter mAdapter = null;
	private List<ColumnData> mList = new ArrayList<>();
	private TextView tvPrompt = null;
	private int page = 1, totalPage = 1;

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_pdf_list, null);
		return view;
	}
	
	@Override
	public void onViewCreated(View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		initListView(view);
		refresh();
	}

	private void refresh() {
		page = 1;
		String url = getArguments().getString(CONST.WEB_URL);
		if (!TextUtils.isEmpty(url)) {
			String[] array = url.split("/");
			url = url.replace("/"+array[array.length-1], "/"+page);
		}
		OkHttpList(url);
	}

	private void onload() {
		page++;
		if (page > totalPage) {//最后一页
			return;
		}
		String url = getArguments().getString(CONST.WEB_URL);
		if (!TextUtils.isEmpty(url)) {
			String[] array = url.split("/");
			url = url.replace("/"+array[array.length-1], "/"+page);
		}
		OkHttpList(url);
	}
	
	private void initListView(View view) {
		tvPrompt = view.findViewById(R.id.tvPrompt);

		listView = view.findViewById(R.id.listView);
		mAdapter = new PDFListAdapter(getActivity(), mList);
		listView.setAdapter(mAdapter);
		listView.setOnItemClickListener(new OnItemClickListener() {
			@Override
			public void onItemClick(AdapterView<?> arg0, View arg1, int arg2, long arg3) {
				ColumnData dto = mList.get(arg2);
				Intent intent = new Intent(getActivity(), PDFActivity.class);
				intent.putExtra(CONST.ACTIVITY_NAME, dto.title);
				intent.putExtra(CONST.WEB_URL, dto.detailUrl);
				startActivity(intent);
			}
		});
		listView.setOnScrollListener(new AbsListView.OnScrollListener() {
			@Override
			public void onScrollStateChanged(AbsListView view, int scrollState) {
                if (scrollState == AbsListView.OnScrollListener.SCROLL_STATE_IDLE && view.getLastVisiblePosition() == view.getCount() - 1) {
					onload();
                }
			}

			@Override
			public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
			}
		});
	}

	private void OkHttpList(final String url) {
		new Thread(new Runnable() {
			@Override
			public void run() {
				OkHttpUtil.enqueue(new Request.Builder().url(url).build(), new Callback() {
					@Override
					public void onFailure(Call call, IOException e) {

					}

					@Override
					public void onResponse(Call call, Response response) throws IOException {
						if (!response.isSuccessful()) {
							return;
						}
						final String result = response.body().string();
						getActivity().runOnUiThread(new Runnable() {
							@Override
							public void run() {
								if (!TextUtils.isEmpty(result)) {
									try {
										JSONObject obj = new JSONObject(result);
										if (!obj.isNull("totalPage")) {
											totalPage = Integer.parseInt(obj.getString("totalPage"));
										}
										if (!obj.isNull("products")) {
											JSONArray array = obj.getJSONArray("products");
											for (int i = 0; i < array.length(); i++) {
												ColumnData dto = new ColumnData();
												JSONObject itemObj = array.getJSONObject(i);
												if (!itemObj.isNull("title")) {
													dto.title = itemObj.getString("title");
												}
												if (!itemObj.isNull("publicTime")) {
													dto.time = itemObj.getString("publicTime");
												}
												if (!itemObj.isNull("filePath")) {
													dto.detailUrl = itemObj.getString("filePath");
												}
												mList.add(dto);
											}

											if (mAdapter != null) {
												mAdapter.notifyDataSetChanged();
											}

											if (mList.size() == 0) {
												tvPrompt.setVisibility(View.VISIBLE);
											}

										}
									} catch (JSONException e) {
										e.printStackTrace();
									}
								}
							}
						});
					}
				});
			}
		}).start();
	}
	
}

File: app/src/main/java/com/cxwl/shawn/wuzhishan/decision/fragment/PdfListFragment.kt
package com.cxwl.shawn.wuzhishan.decision.fragment

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.support.v4.app.Fragment
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AbsListView
import android.widget.AdapterView.OnItemClickListener
import com.cxwl.shawn.wuzhishan.decision.R
import com.cxwl.shawn.wuzhishan.decision.activity.PDFActivity
import com.cxwl.shawn.wuzhishan.decision.activity.WebviewActivity
import com.cxwl.shawn.wuzhishan.decision.adapter.PDFListAdapter
import com.cxwl.shawn.wuzhishan.decision.common.CONST
import com.cxwl.shawn.wuzhishan.decision.dto.ColumnData
import com.cxwl.shawn.wuzhishan.decision.util.OkHttpUtil
import kotlinx.android.synthetic.main.fragment_pdf_list.*
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Request
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException
import java.util.*

/**
 * pdf文档列表
 * @author shawn_sun
 */
class PdfListFragment : Fragment() {

    private var mAdapter: PDFListAdapter? = null
    private val dataList: ArrayList<ColumnData> = ArrayList()
    private var page = 1
    private var totalPage = 1
    private val mUIHandler: Handler = Handler()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_pdf_list, null)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initRefreshLayout()
        initListView()
        refresh()
    }

    /**
     * 初始化下拉刷新布局
     */
    private fun initRefreshLayout() {
        refreshLayout.setColorSchemeResources(CONST.color1, CONST.color2, CONST.color3, CONST.color4)
        refreshLayout.setProgressViewEndTarget(true, 400)
        refreshLayout.isRefreshing = true
        refreshLayout.setOnRefreshListener {
            refresh()
        }
    }

    private fun refresh() {
        page = 1
        var url = arguments!!.getString(CONST.WEB_URL)
        if (!TextUtils.isEmpty(url)) {
            val array = url.split("/")
            url = url.replace("/" + array[array.size - 1], "/$page")
        }
        dataList.clear()
        okHttpList(url)
    }

    private fun onload() {
        page++
        if (page > totalPage) {//最后一页
            return
        }
        var url = arguments!!.getString(CONST.WEB_URL)
        if (!TextUtils.isEmpty(url)) {
            val array = url.split("/")
            url = url.replace("/" + array[array.size - 1], "/$page")
        }
        okHttpList(url)
    }

    private fun initListView() {
        mAdapter = PDFListAdapter(activity!!, dataList)
        listView.adapter = mAdapter
        listView.onItemClickListener = OnItemClickListener { adapterView, view, i, l ->
            val dto = dataList[i]
            val intent = if (dto.detailUrl.endsWith(".pdf") || dto.detailUrl.endsWith(".PDF")) {
                Intent(activity, PDFActivity::class.java)
            } else {
                Intent(activity, WebviewActivity::class.java)
            }
            intent.putExtra(CONST.ACTIVITY_NAME, dto.title)
            intent.putExtra(CONST.WEB_URL, dto.detailUrl)
            startActivity(intent)
        }
        listView.setOnScrollListener(object : AbsListView.OnScrollListener {
            override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {
                if (scrollState == AbsListView.OnScrollListener.SCROLL_STATE_IDLE && view.lastVisiblePosition == view.count - 1) {
                    onload()
                }
            }
            override fun onScroll(view: AbsListView, firstVisibleItem: Int, visibleItemCount: Int, totalItemCount: Int) {}
        })
    }

    private fun okHttpList(url: String) {
        if (TextUtils.isEmpty(url)) {
            tvPrompt.visibility = View.VISIBLE
            refreshLayout.isRefreshing = false
            return
        }
        Thread(Runnable {
            OkHttpUtil.enqueue(Request.Builder().url(url).build(), object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                }

                override fun onResponse(call: Call, response: Response) {
                    if (!response.isSuccessful) {
                        return
                    }
                    val result = response.body!!.string()
                    mUIHandler.post {
                        refreshLayout.isRefreshing = false
                        if (!TextUtils.isEmpty(result)) {
                            val obj = JSONObject(result)
                            if (!obj.isNull("totalPage")) {
                                totalPage = Integer.parseInt(obj.getString("totalPage"));
                            }
                            if (!obj.isNull("products")) {
                                val array = obj.getJSONArray("products")
                                for (i in 0 until array.length()) {
                                    val dto = ColumnData()
                                    val itemObj = array.getJSONObject(i)
                                    if (!itemObj.isNull("title")) {
                                        dto.title = itemObj.getString("title")
                                    }
                                    if (!itemObj.isNull("publicTime")) {
                                        dto.time = itemObj.getString("publicTime")
                                    }
                                    if (!itemObj.isNull("filePath")) {
                                        dto.detailUrl = itemObj.getString("filePath")
                                    }
                                    dataList.add(dto)
                                }

                                if (mAdapter != null) {
                                    mAdapter!!.notifyDataSetChanged()
                                }
                                if (dataList.size == 0) {
                                    tvPrompt.visibility = View.VISIBLE
                                }
                            }
                        }
                    }
                }
            })
        }).start()
    }

}
